### DISCUSSION TITLE
Nested authentication using `Authorization` header.

### DISCUSSION THREAD

**Discussion by mjurkus** [minio/minio#11978] [Q&A] [answered]
I'm running minio behind Google's IAP and I would like to reach `minio` using Minio Python API.
To do that I need to add the `Authorization: Bearer JWT_TOKEN` header. More details in [IAP Authentication docs](https://cloud.google.com/iap/docs/authentication-howto#obtaining_an_oidc_token_for_the_default_service_account).

**Problem 1:**
Unfortunately, Minio Python API overwrites the `Authorization` header here https://github.com/minio/minio-py/blob/master/minio/signer.py#L197

I could work-around (I should not do that, though) that by monkey patching `signer.py` and pre-pending `Bearer JWT_TOKEN` to `authorization`:  
`Bearer JWT_TOKEN, AWS4-HMAC-SHA256 Credential=...., ...`

**Problem 2:**
Unfortunately, you cannot mix Authorization mechanisms with IAP. Everything after the bearer keyword is considered the token string, so the token parsing fails.

Maybe there is another approach I could use for such nested authentication?

**Comment by harshavardhana** [comment, marked as answer]
You cannot @mjurkus

**Comment by mjurkus** [comment]
I'll just add my investigation for future generations.

IAP requires a header `Authorization: Bearer JWT_TOKEN`

Minio also uses `Authorization header to authenticate requests and OVERWIRETES the header that is used to authenticate to IAP. If we monkey-patch the Python SDK to prepend `JWT_TOKEN` to what `minio` generates requests will fail

```Authorization: Bearer IAP_JWT_TOKEN AWS4-HMAC-SHA256, Credential=*REDACTED*/20210406/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*```

Unfortunately, RFC7235 states (and it seems the minio server does not follow it), that the `Authorization` header cannot contain list syntax. Because of that the whole header value is treated as `JWT_TOKEN`, which subsequently fails:

```Response code: 401, Content-Type: text/html; charset=UTF-8, Body: Invalid IAP credentials: Unable to parse JWT```

If we leave only `JWT_TOKEN` such as
```Authorization: Bearer IAP_JWT_TOKEN```

we will be able to get through IAP, but the minio request will fail:

```S3 operation failed; code: AccessDenied, message: Access Denied., resource: /my-bucket, request_id: xxx, host_id: , bucket_name: my-bucket```

So, unless `minio` starts using a different header, for example, `X-App-Auth` or similar - IAP and `minio` won't work together.
